service/resource/deployment/resource.go

```diff
@@ -0,0 +1,294 @@
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	deploymentsToCreate, err := toDeployments(createState)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if len(deploymentsToCreate) != 0 {
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "creating deployments in the Azure API")
+
+		resourceGroupName := key.ClusterID(customObject)
+		deploymentsClient, err := r.getDeploymentsClient()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		for _, deploy := range deploymentsToCreate {
+			r.logger.Log("cluster", key.ClusterID(customObject), "debug", fmt.Sprintf("creating deployment %s", deploy.Name))
+
+			params := make(map[string]interface{}, len(deploy.Parameters))
+			for key, val := range deploy.Parameters {
+				params[key] = struct {
+					Value interface{}
+				}{
+					Value: val,
+				}
+			}
+			deployment := azureresource.Deployment{
+				Properties: &azureresource.DeploymentProperties{
+					Mode:       azureresource.Complete,
+					Parameters: &params,
+					TemplateLink: &azureresource.TemplateLink{
+						URI:            to.StringPtr(deploy.TemplateURI),
+						ContentVersion: to.StringPtr(deploy.TemplateVersion),
+					},
+				},
+			}
+
+			_, errchan := deploymentsClient.CreateOrUpdate(resourceGroupName, deploy.Name, deployment, nil)
+			select {
+			case err := <-errchan:
+				if err != nil {
+					return microerror.Mask(err)
+				}
+			case <-time.After(createTimeout):
+				return microerror.Mask(createTimeoutError)
+			}
+
+			r.logger.Log("cluster", key.ClusterID(customObject), "debug", fmt.Sprintf("created deployment %s", deploy.Name))
+		}
+
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "created the deployments in the Azure API")
+	}
+
+	return nil
+}
+
+// ProcessDeleteState returns nil because deployments are not deleted.
+func (r *Resource) ProcessDeleteState(obj, deleteState interface{}) error {
+	return nil
+}
+
+// ProcessUpdateState is not yet implemented.
+func (r *Resource) ProcessUpdateState(obj, updateState interface{}) error {
+	return nil
+}
+
+// Underlying returns the underlying resource.
+func (r *Resource) Underlying() framework.Resource {
+	return r
+}
+
+func (r *Resource) getDeploymentsClient() (*azureresource.DeploymentsClient, error) {
+	azureClients, err := client.NewAzureClientSet(r.azureConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return azureClients.DeploymentsClient, nil
+}
+
+func existsDeploymentByName(list []Deployment, name string) bool {
+	for _, d := range list {
+		if d.Name == name {
+			return true
+		}
+	}
+
+	return false
+}
+
+func getDeploymentByName(list []Deployment, name string) (Deployment, error) {
+	for _, d := range list {
+		if d.Name == name {
+			return d, nil
+		}
+	}
+
+	return Deployment{}, microerror.Maskf(notFoundError, name)
+}
+
+func toCustomObject(v interface{}) (azuretpr.CustomObject, error) {
+	customObjectPointer, ok := v.(*azuretpr.CustomObject)
```

Comment:
Shouldn't we check for `v` being nil here?

Reply:
Also done.